### PR TITLE
feat(mastracode): v11.6.0 #224 ensureFeatureBranch tool + pre-commit/pre-push guards

### LINKED ISSUE #224: Pipeline can commit to default branch when architect Step 1 is skipped
## Problem

If architect Step 1 (feature-branch creation) is skipped, fails, or is bypassed (e.g. via fast-mode shortcuts), the executor will happily commit straight to `main`. There is no defense-in-depth check before `git commit`.

Adjacent gaps:

- Architect Step 1 instructs the agent to shell out to git for branch creation. There is no validation against local/remote name collisions, no default-branch detection, and no persistence guarantee for `branchName` / `issueNumber` in `luca-state.json`.
- Finalize pushes and opens the PR without re-checking which branch HEAD is on, so a guard regression in execute would surface as a PR opened against `main`.
- The Co-Authored-By trailer points at `noreply@luca.dev`, which doesn't resolve, so GitHub linkback for co-authors is broken.

## Fix

1. New `ensureFeatureBranch` tool with `status` / `create` / `rename` actions. Encapsulates default-branch detection (`origin/HEAD` → `main`/`master`/`trunk` fallback), local + remote collision checks, and persistence of `branchName` + `issueNumber` to `luca-state.json`.
2. Architect Step 1 calls the tool instead of shelling out.
3. Executor adds Step 0 pre-commit guard: `ensureFeatureBranch({ action: "status" })`. Aborts with `BRANCH_NOT_CREATED` on `on-default` / `detached` / `no-git`.
4. Finalize Step 5b.3 adds the same status check before push.
5. Co-Authored-By trailer fixed to `Claude <noreply@anthropic.com>` until `luca.dev` is owned and wired.

## Mode permissions

- architect: `*`
- execute: `["status"]` (read-only self-check)
- finalize: `["status"]`
- build / fast: `*`

### PR DESCRIPTION
**Pull request by asibilia**
Closes #224

## Summary

Hardens feature-branch creation so the pipeline can never silently commit to the default branch. Adds a typed `ensureFeatureBranch` tool that owns branch lifecycle (create / status / rename) with proper default-branch detection, local + remote collision checks, and `luca-state.json` persistence. Wires it into architect (creator), executor (pre-commit guard), and finalize (pre-push guard). Also corrects the Co-Authored-By trailer so GitHub linkback works.

## Why

If architect Step 1 is skipped, fails, or is bypassed (fast-mode shortcuts, manual mode-switch, instruction drift), the executor previously had no defense — it would commit straight onto `main`. This is recoverable but ugly: the orchestrator has to rewind, cherry-pick, and recreate state.

The fix is defense-in-depth: every mode that can mutate git verifies branch state through the same tool, and the tool encapsulates everything that was previously open-coded in agent instructions (default-branch detection, collision checks, state persistence).

### New tool — `packages/luca-mastracode/src/tools/ensure-feature-branch.ts`

Three actions:

| Action   | Side effects                                              | Returns                                                                                                                                                                  |
| -------- | --------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `status` | _none_ (read-only)                                        | `on-feature` \| `on-default` \| `detached` \| `no-git`                                                                                                                    |
| `create` | `git switch -c <type>/<issue>-<slug>` + writes state      | `created` \| `already-on-feature` \| `branch-mismatch` \| `local-collision` \| `remote-collision` \| `on-feature-mismatch-default` \| `detached` \| `no-git` \| `git-error` |
| `rename` | `git branch -m <new>` + writes state                      | `renamed` \| `local-collision` \| `remote-collision` \| `on-default` \| `detached` \| `no-git` \| `git-error`                                                              |

- **Default-branch detection**: prefers `origin/HEAD`, falls back to `main` / `master` / `trunk`.
- **State persistence**: writes `branchName` and `issueNumber` to `.planning/luca-state.json` directly via `writeLucaState` so the architect doesn't need a second `workflowState write` call.
- **Pure helpers** (`slugify`, `buildBranchName`) exported via `__testing` for unit tests; covered by 11 new tests (slugify edge cases, all 7 branch-type prefixes, with/without issue number).

### Wiring

- `tools/index.ts` — exported.
- `tools/tool-manifest.ts` — registered with mode-scoped permissions:

  | Mode      | Allowed actions    | Why                                                       |
  | --------- | ------------------ | --------------------------------------------------------- |
  | architect | `*`                | Creates branches.                                         |
  | execute   | `["status"]`       | Pre-commit self-check; cannot mutate, only read.          |
  | finalize  | `["status"]`       | Pre-push self-check; cannot mutate, only read.            |
  | build     | `*`                | Ad-hoc workflows + `gh-prepare` skill.                    |
  | fast      | `*`                | Same as build.                                            |

### Instructions

- **`instructions/architect.md` Step 1** — replaces the prior shell-out instructions with a documented call to `ensureFeatureBranch({ action: "create", type, issueNumber, slug })` and an enumeration of return statuses. Removes the trailing redundant `workflowState write` since the tool persists.
- **`instructions/finalize.md` Step 5b.3** — adds a pre-push status check; renumbers subsequent items.

### Subagent

- **`subagents/executor.ts` Step 0** — pre-commit guard. Calls `ensureFeatureBranch({ action: "status" })` once before the first commit. Aborts with `BRANCH_NOT_CREATED` on `on-default` / `detached` / `no-git`. Replaces the shell-out version that landed in the prior session — same contract, but now backed by the tool's encapsulated default-branch detection.

### Co-Authored-By trailer

- Trailer in executor instructions changed from `Luca <noreply@luca.dev>` (domain doesn't resolve, GitHub linkback broken) to `Claude <noreply@anthropic.com>`. Will revisit when `luca.dev` is owned and wired.

## What this catches

| Regression                                              | Caught by                                |
| ------------------------------------------------------- | ---------------------------------------- |
| Architect Step 1 silently skipped                       | Executor `BRANCH_NOT_CREATED` Step 0     |
| User invokes execute on `main` outside the pipeline     | Same executor guard                      |
| Fast / TRIVIAL paths bypass architect                   | Same executor guard (granted to `fast`)  |
| Architect tries to create branch when already on `main` | `local-collision` from `create`          |
| Branch already exists on origin                         | `remote-collision` from `create`         |
| Finalize tries to open PR against `main`                | Finalize pre-push status guard           |
| Detached-HEAD state (mid-rebase)                        | `detached` status from any action        |

## Verification

- `bun run typecheck` — clean.
- `bun test` — 110/110 pass (11 new in `ensure-feature-branch.test.ts`).
- `eslint` on changed files — 0 issues.

## Release

- Changeset: `.changeset/ensure-feature-branch.md` — **`minor`** bump on `@alecsibilia/luca-mastracode` (11.5.0 → 11.6.0).
- Milestone: v11.6.0.

### CODE REVIEW

**Comment by asibilia** [comment]
@copilot please review this PR

File: packages/luca-mastracode/src/tools/ensure-feature-branch.ts [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+                    status: 'invalid-args' as const,
+                    currentBranch: current,
+                    defaultBranch: def,
+                    message:
+                        'action="create" requires both `type` and `slug`. Optional: `issueNumber`.',
+                }
+            }
+
+            // Already on a feature branch and not forcing → idempotent no-op.
+            if (current !== '' && current !== def && !force) {
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/tools/ensure-feature-branch.ts:267]: Fixed in c423697f5 (and earlier 6c79... — merged into the bug-fix commit). The `create` action now performs an explicit detached check before any branch operations and returns `status: 'detached'` with `ok: false`, matching the documented stop-condition taxonomy. Test coverage in `ensure-feature-branch.test.ts` guards the four documented status codes (`detached` | `on-default` | `local-collision` | `remote-collision`).

File: packages/luca-mastracode/src/tools/ensure-feature-branch.ts [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+                    currentBranch: current,
+                    defaultBranch: def,
+                    proposedBranch: target,
+                    message: `Remote branch '${target}' already exists on origin. Pick a different slug or check out the existing branch manually.`,
+                }
+            }
+
+            // Create from default branch (or from current HEAD if current
+            // is the default — same thing).
+            const switched = git(['switch', '-c', target])
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/tools/ensure-feature-branch.ts:337]: Fixed. The `create` action now switches to the default branch _before_ running `git switch -c <target>` whenever `current !== def` (i.e. when `force=true` is used on a non-default feature branch). The new branch is now always created from a clean default-branch base, matching the documented contract. Source-level test guard in `ensure-feature-branch.test.ts` (`creates from default branch when off-default`).

File: packages/luca-mastracode/src/instructions/architect.md [hunk 3]
```diff
@@ -32,11 +32,25 @@ You are **Luca's architect agent**. Create detailed, reviewable execution plans
...
+   ```
+
+   Returned `status` values:
+   - `"created"` — switched to new branch from default. Proceed.
+   - `"already-on-feature"` — repo was already on a feature branch matching the recorded state. Proceed.
+   - `"branch-mismatch"` / `"local-collision"` / `"remote-collision"` — STOP. Report to user; do not retry blindly.
+   - `"on-default"` (only seen via action="status") — must call `action: "create"` first.
+   - `"detached"` / `"no-git"` / `"git-error"` — STOP. Report to user.
+
+3. The tool already persists `branchName` and `issueNumber` to `workflow_state`. No second `workflowState write` is required for those fields.
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/instructions/architect.md]: Fixed wording in c423697f5. Step 3 now reads: _"The tool already persists `branchName` and `issueNumber` to `.planning/luca-state.json` (via `writeLucaState`)."_ Also added explicit `--skip-branch` handling that documents writing `skipBranch: true` to state via `workflowState write` so the executor's pre-commit guard knows to bypass.

File: packages/luca-mastracode/src/subagents/executor.ts [hunk 4]
```diff
@@ -9,11 +9,38 @@ export const executorSubagent: HarnessSubagent = {
...
-5. Stage and commit with a descriptive message
+5. **Pre-commit MuninnDB recall** — before staging, query MuninnDB for prior learnings that could change *what* gets committed (commit-message conventions, sign-off trailers, scope rules, files we've previously committed by mistake). Vault from \`.planning/config.json\` → \`muninn.vault\`, fallback \`"default"\`:
+
+   \`\`\`
+   mcp__muninn__muninn_recall({
+     vault: "<repo_vault>",
+     context: ["commit conventions", "pre-commit pitfalls", "<scope of this task>"],
+     mode: "semantic",
+     limit: 5,
+   })
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/subagents/executor.ts]: Fixed. Switched the snippet from the JS object-literal form `mcp__muninn__muninn_recall({ ... })` to the named-arg form `mcp__muninn__muninn_recall(vault: "<repo_vault>", context: [...], ...)` used by every other recall example in `instructions/*.md`. The same fix was already shipped in `instructions/finalize.md` Step 5b — this one in `executor.ts` Step 5 was the only outlier.

File: packages/luca-mastracode/src/tools/ensure-feature-branch.ts [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+            message: `Unknown action: ${action}`,
+        }
+    },
+})
+
+// Exported for testing.
+export const __testing = {
+    slugify,
+    buildBranchName,
+}
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/tools/ensure-feature-branch.ts:462]: Both halves addressed:

1. **Tests do exist** — `packages/luca-mastracode/src/__tests__/ensure-feature-branch.test.ts` has 12 test cases for `slugify` / `buildBranchName` plus 6 new regression guards I added in this iteration (status taxonomy, isInsideGitRepo stdout validation, create-from-default contract, branchExistsRemote timeout). They were invisible to your review because…

2. **The `__tests__/` directory was gitignored** — root `.gitignore` had a wholesale `__tests__/` + `*.test.ts` exclusion tied to a since-retired `no-tests.rule.ts`. The exclusion silently hid all test files from PR diffs. I removed the exclusion and explicitly committed the test files. They're now visible in the diff. `bun test` reports 116 pass / 0 fail.

File: packages/luca-mastracode/src/tools/ensure-feature-branch.ts [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+            stderr:
+                e.stderr?.toString().trim() ??
+                e.message ??
+                'git command failed',
+        }
+    }
+}
+
+function isInsideGitRepo(): boolean {
+    return git(['rev-parse', '--is-inside-work-tree']).ok
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/tools/ensure-feature-branch.ts]: Fixed. `isInsideGitRepo()` now requires both `r.ok` _and_ `r.stdout === 'true'`. This correctly rejects `.git/` directories and bare repos where `git rev-parse --is-inside-work-tree` exits 0 while printing `false`. Source-level test guard added.

File: packages/luca-mastracode/src/tools/ensure-feature-branch.ts [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+            }
+            if (branchExistsLocal(target) || branchExistsRemote(target)) {
+                return {
+                    ok: false as const,
+                    status: 'collision' as const,
+                    currentBranch: current,
+                    defaultBranch: def,
+                    proposedBranch: target,
+                    message: `Cannot rename: '${target}' already exists locally or on origin.`,
+                }
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/tools/ensure-feature-branch.ts:423]: Fixed. The `rename` action now returns the documented status taxonomy:

- `detached` — was collapsed into `cannot-rename`
- `on-default` — was collapsed into `cannot-rename`
- `local-collision` — was collapsed into `collision`
- `remote-collision` — was collapsed into `collision`

This matches the codes already returned by `create` and `status`, so callers can branch on specific failure modes uniformly across all three actions. Test guard in `ensure-feature-branch.test.ts` asserts the source emits all four codes and never the old collapsed ones.

**Comment by asibilia** [comment]
@copilot — thanks for the thorough review. All 7 inline comments and both medium/low issues from your summary are addressed in the two new commits:

**`fix(mastracode): harden ensureFeatureBranch + un-ignore test files`** (`6c79...` — see `git log`)

| Finding | Resolution |
| --- | --- |
| `create` proceeds on detached HEAD | Explicit `status: 'detached'` reject before branching |
| `create` branches from current HEAD when `force=true` off-default | Switches to default first, then `git switch -c` |
| `branchExistsRemote` no timeout (75s hang risk) | 5 s timeout via `execFileSync({ timeout })`, fail-open (local check still fires) |
| `rename` collapses `cannot-rename` / `collision` | Returns documented `detached` / `on-default` / `local-collision` / `remote-collision` |
| `isInsideGitRepo` only checks exit code | Now requires `r.stdout === 'true'` |
| "11 tests claimed but no tests in repo" | Tests exist; `__tests__/` was gitignored. Removed exclusion, committed test files, added 6 new regression guards. 116 / 0. |

**`fix(mastracode): skipBranch state + doc/syntax fixes from PR review`** (`c423697f5`)

| Finding | Resolution |
| --- | --- |
| `--skip-branch` + executor guard regression | Added `skipBranch?: boolean` to `lucaStateSchema`; architect persists when `--skip-branch` is active; executor guard reads state and bypasses `ensureFeatureBranch` when set |
| `architect.md` says `workflow_state` but code writes `luca-state.json` | Wording corrected |
| Recall syntax mismatch in `executor.ts` Step 5 | Switched from `({ ... })` to named-arg `(vault: ..., context: [...], ...)` matching every other recall example |

**Minor nits — noted, not addressed in this PR:**
- Step 0 numbering collision in `execute.md` vs `executor.ts` — agreed, will tackle in a follow-up consistency pass.
- `gh-prepare/SKILL.md` still uses `git checkout -b` directly — agreed, but the standalone skill path is intentionally git-only (not always in a Luca-state context). A future pass can delegate to `ensureFeatureBranch` when state is present.

Final check: typecheck clean, `bun test` 116 / 0, lint clean for edited files.
